storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +65,387 @@
-            Set<Integer> updatedAssignedMetaPartitions = new HashSet<>();
-            for (TopicIdPartition tp : updatedReassignedPartitions) {
-                updatedAssignedMetaPartitions.add(topicPartitioner.metadataPartition(tp));
-            }
+    public Set<Integer> metadataPartitionsAssigned() {
+        return Collections.unmodifiableSet(assignedMetadataPartitions);
+    }
+
+    private void fetchBeginAndEndOffsets() {
+        try {
+            final Set<TopicPartition> unInitializedPartitions = assignedUserTopicIdPartitions.values().stream()
+                .filter(utp -> utp.isAssigned && !utp.isInitialized)
+                .map(utp -> toRemoteLogPartition(utp.metadataPartition))
+                .collect(Collectors.toSet());
+            // Removing the previous offset holder if it exists. During reassignment, if the list-offset
+            // call to `earliest` and `latest` offset fails, then we should not use the previous values.
+            unInitializedPartitions.forEach(x -> offsetHolderByMetadataPartition.remove(x));
+            if (!unInitializedPartitions.isEmpty()) {
+                Map<TopicPartition, Long> endOffsets = consumer.endOffsets(unInitializedPartitions);
+                Map<TopicPartition, Long> beginOffsets = consumer.beginningOffsets(unInitializedPartitions);
+                offsetHolderByMetadataPartition = endOffsets.entrySet()
+                    .stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey,
+                        e -> new BeginAndEndOffsetHolder(beginOffsets.get(e.getKey()), e.getValue())));
 
-            // Clear removed topic partitions from in-memory cache.
-            for (TopicIdPartition removedPartition : removedPartitions) {
-                remotePartitionMetadataEventHandler.clearTopicPartition(removedPartition);
             }
+            isOffsetsFetchFailed = false;
+        } catch (final RetriableException ex) {
+            // ignore LEADER_NOT_AVAILABLE error, this can happen when the partition leader is not yet assigned.
+            isOffsetsFetchFailed = true;
+            lastFailedFetchOffsetsTimestamp = time.milliseconds();
+        }
+    }
 
-            assignedTopicPartitions = Collections.unmodifiableSet(updatedReassignedPartitions);
-            log.debug("Assigned topic partitions: {}", assignedTopicPartitions);
+    private void maybeFetchBeginAndEndOffsets() {
+        // If the leader for a `__remote_log_metadata` partition is not available, then the call to `ListOffsets`
+        // will fail after the default timeout of 1 min. Added a delay of 5 min in between the retries to prevent the
+        // thread from aggressively fetching the list offsets. During this time, the recently reassigned
+        // user-topic-partitions won't be marked as initialized.
+        if (isOffsetsFetchFailed && lastFailedFetchOffsetsTimestamp + 300_000 < time.milliseconds()) {
+            fetchBeginAndEndOffsets();
+        }
```

Comment:
The 5 mins wait is too long. Like you said in the comment, timeout is 1 min, why can't we retry after 10 secs or so  to have more chance to get the offsets? IMO, 10 secs for electing a leader for a partition is long enough, so we should reduce this timeout. Thought?

Reply:
When the leader is offline, it may take some time to come up. In that case, we want to avoid aggressively fetching the offsets. Otherwise, we will frequently call this method and get blocked for 1 min (until the API times out). This will prevent us from processing updates for the already assigned topic partitions. 

It is basically a tradeoff (when the leader is offline) between whether we want to update for the newly assigned topic partitions as soon as possible or always be up-to-date for the already assigned topic partitions. We chose the second option here.

Reply:
Added input parameter